server/src/graph_maker/tram_track_graph_transformer.py

```diff
@@ -0,0 +1,386 @@
+import math
+
+import networkx as nx
+import overpy
+import pyproj
+from shapely.geometry import LineString, MultiLineString, Point
+from shapely.ops import linemerge, transform
+
+
+class TramTrackGraphTransformer:
+    """
+    TramTrackGraphTransformer converts OSM data into a directed NetworkX graph in two main stages:
+
+        1. Skeleton graph construction:
+        - Builds an initial graph (_graph) using only main nodes — tram stops and tram track intersections.
+
+        2. Graph densification:
+        - Using the `densify_graph_by_max_distance` method, the graph is densified by adding  intermediate nodes so that the distance between consecutive nodes does not exceed the specified threshold.
+    """
+
+    def __init__(self, tram_stops_and_tracks: overpy.Result):
+        self._ways = tram_stops_and_tracks.ways
+        self._stops = [
+            node
+            for node in tram_stops_and_tracks.nodes
+            if node.tags.get("railway") == "tram_stop"
+        ]
+        self._coords_data = tram_stops_and_tracks.nodes
+        self._permanent_nodes = set()
+        self._ways_dict = {}
+
+        self._initialize_graph()
+
+    def _build_initial_graph(self):
+        """
+        Builds a directed graph from tram track ways.
+        Only forward edges are added; reverse edges (for two-way tracks)
+        are handled later in a separate method.
+        """
+        graph = nx.DiGraph()
+
+        for way in self._ways:
+            node_ids = [node.id for node in way.nodes]
+
+            for i in range(len(node_ids) - 1):
+                graph.add_edge(node_ids[i], node_ids[i + 1])
+
+        return graph
+
+    def _find_permanent_tram_stops(self):
+        result = set(
+            stop.id
+            for stop in self._stops
+            if stop.id in self._graph.nodes
+            and (
+                self._graph.out_degree(stop.id) > 0
+                or self._graph.in_degree(stop.id) > 0
+            )
+        )
+
+        return result
```

Comment:
You can use a set comprehension here. Also there's no need to assign it to variable `result` as it's returned directly in the next instruction.

```suggestion
        return {
            stop.id
            for stop in self._stops
            if stop.id in self._graph.nodes
            and (
                self._graph.out_degree(stop.id) > 0
                or self._graph.in_degree(stop.id) > 0
            )
        }
```